File: mindflow/utils/git.py
"""
Handles git related operations.
"""
import logging
import os
import subprocess
from typing import List


def is_within_git_repo(path: os.PathLike) -> bool:
    """
    Checks if the given path is within a git repository.
    """
    try:
        output = subprocess.run(
            ["git", "-C", path, "rev-parse", "--git-dir"],
            capture_output=True,
            timeout=10,
            check=True,
        )
        return output.returncode == 0
    except Exception:
        return False


def get_git_files(path: os.PathLike) -> List[str]:
    """
    Extract all files from a git repository that are not gitignored.
    """
    try:
        output = subprocess.run(
            ["git", "-C", os.fspath(path), "ls-files"],
            capture_output=True,
            timeout=10,
            check=True,
        )
        git_files = output.stdout.decode().strip().split("\n")
        return [os.path.join(os.fspath(path), f) for f in git_files]
    except Exception as error:
        logging.debug("Failed extract git files with 'git ls-files': %s", error)
        return []

File: mindflow/src/utils/git.rs
use std::path::Path;
use std::process::Command;
use std::string::String;

pub fn is_within_git_repo(path: &Path) -> bool {
    let output = Command::new("git")
                .arg("-C")
                .arg(path.to_string_lossy().to_string())
                .arg("rev-parse")
                .arg("--git-dir")
                .output()
                .expect("failed to execute process");
    if output.status.success() {
        true
    } else {
        false
    }
}

pub fn is_git_repo_head(path: &Path) -> bool {
    let git_dir = path.join(".git");
    git_dir.exists() && git_dir.is_dir()
}

pub fn get_git_files(path: &Path) -> Vec<String> {
    let output = Command::new("git")
                .arg("-C")
                .arg(path.to_string_lossy().to_string())
                .arg("ls-files")
                .output()
                .expect("failed to execute process");

    let git_files = String::from_utf8(output.stdout).expect("invalid utf8");
    return git_files.lines().map(|s| format!("{}/{}", path.to_string_lossy(), s.to_string())).collect();
}
